www/js/plugin/storage.ts

```diff
@@ -0,0 +1,182 @@
+import { getAngularService } from "../angular-react-helper";
+import { addStatReading, statKeys } from "./clientStats";
+import { displayErrorMsg, logDebug } from "./logger";
+
+const mungeValue = (key, value) => {
+  let store_val = value;
+  if (typeof value != "object") {
+    // Should this be {"value": value} or {key: value}?
+    store_val = {};
+    store_val[key] = value;
+  }
+  return store_val;
+}
+
+/*
+ * If a non-JSON object was munged for storage, unwrap it.
+ */
+const unmungeValue = (key, retData) => {
+  if (retData?.[key]) {
+    // it must have been a simple data type that we munged upfront
+    return retData[key];
+  } else {
+    // it must have been an object
+    return retData;
+  }
+}
+
+const localStorageSet = (key: string, value: {[k: string]: any}) => {
+  localStorage.setItem(key, JSON.stringify(value));
+}
+
+const localStorageGet = (key: string) => {
+  const value = localStorage.getItem(key);
+  if (value) {
+    return JSON.parse(value);
+  } else {
+    return null;
+  }
+}
+
+/* We redundantly store data in both local and native storage. This function checks
+    both for a value. If a value is present in only one, it copies it to the other and returns it.
+  If a value is present in both, but they are different, it copies the native value to
+    local storage and returns it. */
+function getUnifiedValue(key) {
+  let ls_stored_val = localStorageGet(key);
+  return window['cordova'].plugins.BEMUserCache.getLocalStorage(key, false).then((uc_stored_val) => {
+    logDebug(`for key ${key}, uc_stored_val = ${JSON.stringify(uc_stored_val)},
+                                  ls_stored_val = ${JSON.stringify(ls_stored_val)}.`);
+
+    /* compare stored values by stringified JSON equality, not by == or ===.
+      for objects, == or === only compares the references, not the contents of the objects */
+    if (JSON.stringify(ls_stored_val) == JSON.stringify(uc_stored_val)) {
+      logDebug("local and native values match, already synced");
+      return uc_stored_val;
+    } else {
+      // the values are different
+      if (ls_stored_val == null) {
+        console.assert(uc_stored_val != null, "uc_stored_val should be non-null");
+        logDebug(`for key ${key}, uc_stored_val = ${JSON.stringify(uc_stored_val)},
+                                      ls_stored_val = ${JSON.stringify(ls_stored_val)}.
+                                      copying native ${key} to local...`);
+        localStorageSet(key, uc_stored_val);
+        return uc_stored_val;
+      } else if (uc_stored_val == null) {
+        console.assert(ls_stored_val != null);
+        /*
+         * Backwards compatibility ONLY. Right after the first
+         * update to this version, we may have a local value that
+         * is not a JSON object. In that case, we want to munge it
+         * before storage. Remove this after a few releases.
+         */
```

Comment:
we should remove this comment since it has already "been a few releases", and replace it with one that indicates that we want to copy the value from the local storage to the plugin storage.
- Not sure if we want to display the error - it is apparently not that infrequent.
- I don't think we need to munge the value any more since that was a backwards compat fix - we now just want to duplicate the value between local storage and the plugin storage.

Reply:
What do you think about replacing `displayErrorMsg` with `logWarn` here?

`logWarn` is not user-facing so it won't be annoying, but if we need to check logs we should be able to find it more easily since it's a higher log level than `DEBUG` or `INFO`

Reply:
> * I don't think we need to munge the value any more since that was a backwards compat fix

You're just talking about line 73 right? We still need to munge/unmunge everywhere else?

Reply:
removed the backwards compat